File: src/uva/Leetcode00017.swift
class Solution {
    func combo(digits : [String], _ pad : [String:String], _ i : Int, _ s : String, inout _ c : [String]) {
        if i == digits.count {
            c.append(s)
            return
        }
        for p in pad[digits[i]]!.characters {
            combo(digits, pad, i+1, s + String(p), &c)
        }
    }

    func letterCombinations(digits: String) -> [String] {
        var pad = [
            "2":"abc",
            "3":"def",
            "4":"ghi",
            "5":"jkl",
            "6":"mno",
            "7":"pqrs",
            "8":"tvu",
            "9":"wxyz",
        ]
        if digits.characters.count == 0 {
            return []
        }
        var c = [String]()
        combo(digits.characters.map{String($0)}, pad, 0, "", &c)
        return c
    }
}

File: src/main/Leetcode00017.java
import java.util.*;

public class Leetcode00017 {
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return new ArrayList<>(0);
        }
        String map[] = new String[]{
            "",
            "",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz",
        };
        List<String> l = new ArrayList<>();
        combine(l, "", digits, 0, map);
        return l;
    }
    
    public void combine(List<String> list, String desc, String src, int i, String[] m) {
        if (src.length() == i) {
            list.add(desc);
        } else {
            int digit = src.charAt(i) - '0';
            for (int j = 0;j < m[digit].length();j++) {
                combine(list, desc + m[digit].charAt(j), src, i+1, m);
            }
        }
    }
}
